src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5149,7 +5149,27 @@ public override BoundNode VisitUnaryOperator(BoundUnaryOperator node)
         {
             Debug.Assert(!IsConditionalState);
 
-            _ = base.VisitUnaryOperator(node);
+            switch (node.OperatorKind)
+            {
+                case UnaryOperatorKind.BoolLogicalNegation:
+                    VisitCondition(node.Operand);
+                    SetConditionalState(StateWhenFalse, StateWhenTrue);
+                    break;
+                case UnaryOperatorKind.DynamicTrue:
+                    // We cannot use VisitCondition, because the operand is not of type bool.
+                    // Yet we want to keep the result split if it was split.  So we simply visit.
+                    Visit(node.Operand);
+                    break;
+                case UnaryOperatorKind.DynamicFalse:
+                    Visit(node.Operand);
+                    Split();
+                    SetConditionalState(StateWhenFalse, StateWhenTrue);
```

Comment:
>SetConditionalState(StateWhenFalse, StateWhenTrue); [](start = 20, length = 51)

Instead of always splitting the state, can we do this conditionally, only when we are in a conditional state (```IsConditionalState```)?

Reply:
As a rule we have avoided such micro-optimizations that complicate the code for little expected benefit.  The caller is likely to split the state in any case.

---
In reply to: [276385502](https://github.com/dotnet/roslyn/pull/34996#discussion_r276385502) [](ancestors = 276385502)